scripts/calculateRevenue/protocols/beefy/beefy.ts

```diff
@@ -0,0 +1,112 @@
+import {
+  BeefyInvestorTransactionWithUsdBalance,
+  BeefyVaultTvlData,
+  VaultInfo,
+} from './types'
+import { getStrategyContract } from '../utils/viem'
+import { RevenueResult, NetworkId } from '../../../types'
+import { getVaults } from './getVaults'
+
+/**
+ * Given a list of transaction history on a particular Beefy vault and a timestamp,
+ * returns a user's TVL in the vault at the specified time.
+ */
+export function getUserTvl(
+  txHistory: BeefyInvestorTransactionWithUsdBalance[],
+  timestamp: Date,
+): number {
+  let lastTvl = txHistory[0].usd_balance
+  for (const tx of txHistory) {
+    if (new Date(tx.datetime) >= timestamp) {
+      return lastTvl
+    }
+    lastTvl = tx.usd_balance
+  }
+  return lastTvl
+}
+
+/**
+ * Given a list of Beefy vault TVL data and a timestamp, returns the vault's TVL at
+ * the specified time.
+ */
+export function getVaultTvl(
+  tvlHistory: BeefyVaultTvlData[],
+  timestamp: Date,
+): number {
+  let lastTvl = tvlHistory[0][1]
+  for (const tvl of tvlHistory) {
+    if (new Date(tvl[0]) >= timestamp) {
+      return lastTvl
+    }
+    lastTvl = tvl[1]
+  }
+  return lastTvl
+}
+
+/**
+ * Given historical information about a vault and a user's transactions, calculates the amount of fee
+ * revenue generated by the user's interaction with the vault, denominated in the native currency
+ * of the chain the vault is deployed on.
+ */
+export async function calculateVaultRevenue(vaultInfo: VaultInfo): Promise<{
+  tokenId: string
+  revenue: string
+}> {
+  let totalRevenueContribution = 0n
+  for (const { beefyFee, timestamp } of vaultInfo.feeEvents) {
+    const userTvl = getUserTvl(vaultInfo.txHistory, timestamp)
+    const vaultTvl = getVaultTvl(vaultInfo.vaultTvlHistory, timestamp)
+    const partialRevenueContribution =
+      (BigInt(userTvl * 10 ** 18) * BigInt(beefyFee)) /
+      BigInt(vaultTvl * 10 ** 18)
```

Comment:
nit: should we do the 10 ** 18 multiplication after converting to bigint to be safe?

Reply:
We have to do the conversion before; converting the values to BigInt directly doesn't work since they have lots of decimal precision, and rounding would lose all of that.